### PR TITLE
[WIP] Add Text Classification TensorFlow Tutorial

### PR DESCRIPTION
**Pull request by JRAlexander**
Add Text Classification TensorFlow Tutorial.

[Related Code Sample](https://github.com/dotnet/samples/pull/935)
[Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/tutorials/text-classification-tf?branch=pr-en-us-12587)
[AB#1530497](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1530497)

### CODE REVIEW

**Comment by JRAlexander** [comment]
closing and reopening

**Comment by JRAlexander** [comment]
Closing and reopening

File: docs/machine-learning/tutorials/text-classificationTF.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    - In the **Add New Item** dialog box, select **Class** and change the **Name** field to *IMDBData.cs*. Then, select the **Add** button.
+
+5. The *IMDBData.cs* file opens in the code editor. Add the following `using` statement to the top of *IMDBData.cs*:
+
+   [!code-csharp[AddUsings](../../../samples/machine-learning/tutorials/TextClassificationTF/IMDBData.cs#AddUsings "Add necessary usings")]
+
+6. Remove the existing class definition and add the following code, which has two classes `IMDBSentiment` and `IMDBPrediction`, to the *IMDBData.cs* file:
+
+   [!code-csharp[DeclareTypes](../../../samples/machine-learning/tutorials/TextClassificationTF/IMDBData.cs#DeclareTypes "Declare data record types")]
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classificationTF.md]: This is showing up empty in the rendering

File: docs/machine-learning/tutorials/text-classificationTF.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+### How the data was prepared
+
+The input dataset class, `IMDBSentiment`, has a `string` for user comments (`SentimentText`) and an `integer` array (`VariableLengthFeatures`)  In addition, the `VariableLengthFeatures` property has a [VectorType](xref:Microsoft.ML.Data.VectorTypeAttribute.%23ctor%2A) attribute to designate the vector type.
+
+`IMDBPrediction` is the prediction class used after the model training. `SentimentPrediction` has a single `float` array (`Prediction`) and a `VectorType` attribute.
+
+The [MLContext class](xref:Microsoft.ML.MLContext) is a starting point for all ML.NET operations. Initializing `mlContext` creates a new ML.NET environment that can be shared across the model creation workflow objects. It's similar, conceptually, to `DBContext` in Entity Framework.
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classificationTF.md]: This is not related to "How the data was prepared"

File: docs/machine-learning/tutorials/text-classificationTF.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+   [!code-csharp[AddUsings](../../../samples/machine-learning/tutorials/TextClassificationTF/IMDBData.cs#AddUsings "Add necessary usings")]
+
+6. Remove the existing class definition and add the following code, which has two classes `IMDBSentiment` and `IMDBPrediction`, to the *IMDBData.cs* file:
+
+   [!code-csharp[DeclareTypes](../../../samples/machine-learning/tutorials/TextClassificationTF/IMDBData.cs#DeclareTypes "Declare data record types")]
+
+### How the data was prepared
+
+The input dataset class, `IMDBSentiment`, has a `string` for user comments (`SentimentText`) and an `integer` array (`VariableLengthFeatures`)  In addition, the `VariableLengthFeatures` property has a [VectorType](xref:Microsoft.ML.Data.VectorTypeAttribute.%23ctor%2A) attribute to designate the vector type.
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classificationTF.md]: Suggest adding a preview of the data to make these explanations more accessible

**Review by natke — COMMENTED** [review]
I started to review this but most of the code blocks are blank. Is the code under review, and do I need to review it? Will finish this review once we get the code sorted.

**Comment by JRAlexander** [comment]
closing and reopening for samples code updates

**Comment by JRAlexander** [comment]
closing and reopening to retrigger build

**Review by luisquintanilla — COMMENTED** [review]
Few comments

**Comment by JRAlexander** [comment]
closing and reopening

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+5. The *IMDBData.cs* file opens in the code editor. Add the following `using` statement to the top of *IMDBData.cs*:
+
+   [!code-csharp[AddUsings](../../../samples/machine-learning/tutorials/TextClassificationTF/IMDBData.cs#AddUsings "Add necessary usings")]
+
+6. Remove the existing class definition and add the following code, which has two classes `IMDBSentiment` and `IMDBPrediction`, to the *IMDBData.cs* file:
+
+   [!code-csharp[DeclareTypes](../../../samples/machine-learning/tutorials/TextClassificationTF/IMDBData.cs#DeclareTypes "Declare data record types")]
+
+### How the data was prepared
```
[the remark this answers is not in the document]
Comment (natke) [reply docs/machine-learning/tutorials/text-classification-tf.md]: A preview of the data would be helpful here too

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+`IMDBPrediction` is the prediction class used after the model training. `IMDBPrediction` has a single `float` array (`Prediction`) and a `VectorType` attribute.
+
+### Adding context
+
+The [MLContext class](xref:Microsoft.ML.MLContext) is a starting point for all ML.NET operations. Initializing `mlContext` creates a new ML.NET environment that can be shared across the model creation workflow objects. It's similar, conceptually, to `DBContext` in Entity Framework.
+
+1. Replace the `Console.WriteLine("Hello World!")` line in the `Main` method with the following code to declare and initialize the mlContext variable:
+
+   [!code-csharp[CreateMLContext](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#CreateMLContext "Create the ML Context")]
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: Why are we using the seed of 1 here? Do we need the seed? We probably should explain it if we need it

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+1. Add the following as the next line of code in the `Main()` method:
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: This code won't compile now. Some people seem to find that really annoying. Can we refactor so that the code is always compilable throughout the tutorial?

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+   [!code-csharp[CallReuseAndTuneSentimentModel](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#CallReuseAndTuneSentimentModel)]
+
+### Create an intermediate features class
+
+1. Create a utility class called `IntermediateFeatures` to hold intermediate data that will be used by the CustomMapping Estimator in a later step. Add this after the `Main()` method:
+
+[!code-csharp[DeclareIntermediateFeatures](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#DeclareIntermediateFeatures)]
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: Code need indenting

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    * Tunes (retrains) the model.
+    * Returns the model.
+
+1. Create an instance of `IMDBSentiment` called `trainData` and pass it to the `Prediction Engine` by adding the following as the next lines of code in the `ReuseAndTuneSentimentModel()` method:
+
+   [!code-csharp[CreateTrainData](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#CreateTrainData)]
+
+### Load the data
+
+Data in ML.NET is represented as an [IDataView class](xref:Microsoft.ML.IDataView). `IDataView` is a flexible, efficient way of describing tabular data (numeric and text). Data can be loaded from a text file or in real time (for example, SQL database or log files) to an `IDataView` object.
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: "Other sources" is  better than "real time" as these other sources are necessarily real time

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+1. Load the in-memory collection into an [`IDataView`](xref:Microsoft.ML.IDataView) with the [`LoadFromEnumerable`](xref:Microsoft.ML.DataOperationsCatalog.LoadFromEnumerable%2A) method:
+
+   [!code-csharp[LoadTrainData](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#LoadTrainData)]
+
+### Extract and transform the data
+
+1. Create a dictionary to convert words into the integer indexes using the [`LoadFromTextFile`](xref:Microsoft.ML.TextLoaderSaverCatalog.LoadFromTextFile%2A) method to load data from a file, as seen in the following table:
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: It's not clear to me why we are performing this step. Can we add a bit more explanation?

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+1. This code will function as a lookup map to assist with mapping text to integer vectors in later steps, so add it next to the `ReuseAndTuneSentimentModel` method:
+
+   [!code-csharp[CreateLookupMap](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#CreateLookupMap)]
+
+1. Append the `TensorFlowTransform` to the `pipeline` with the following line of code:
+
+   [!code-csharp[LoadTensorFlowModel](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#LoadTensorFlowModel)]
+
+   The `LoadTensorFlowModel` is a convenience method that allows the `TensorFlow` model to be loaded once and then creates the `TensorFlowEstimator` using `ScoreTensorFlowModel`. `Prediction` returns a probability for sentiment of a given text, and all of those probabilities must add up to 1.
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: Not clear why we are talking about `Prediction` here.

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+    The schema is output to the console:
+
+    ```console
+    =============== TensorFlow Model Schema ===============
+    Name: Features, Type: System.Int32, Shape: (-1, 600)
+    Name: Prediction/Softmax, Type: System.Int32, Shape: (-1, 600)
+    ```
+
+1. Add an `Action` to resize the features as input to a `CustomMapping` transform with the next lines of code:
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: Again, why are we doing this?

File: docs/machine-learning/tutorials/text-classification-tf.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+[!code-csharp[CopyColumns](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#CopyColumns)]
+
+The `Fit()` method trains your model by transforming the dataset and applying the training. Fit the model to the training dataset and return the trained model by adding the following as the next line of code in the `ReuseAndTuneSentimentModel()` method:
+
+[!code-csharp[TrainModel](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#TrainModel)]
+
+Return the trained model by adding the following as the last line of code in the `ReuseAndTuneSentimentModel()` method:
+
+[!code-csharp[ReturnModel](~/samples/machine-learning/tutorials/TextClassificationTF/Program.cs#ReturnModel)]
```
Comment (natke) [inline docs/machine-learning/tutorials/text-classification-tf.md]: This whole pipeline section could do with a diagram. Just basic boxes would suffice

**Review by natke — COMMENTED** [review]
A few comments from me
